stdlib/string.ml

```diff
@@ -98,20 +100,43 @@ let escaped s =
   else
     s
 
-let index s c =
-  B.index (bos s) c
-let rindex s c =
-  B.rindex (bos s) c
-let index_from s i c=
-  B.index_from (bos s) i c
+let rec index_rec s lim i c =
+  if i >= lim then raise Not_found else
+  if unsafe_get s i = c then i else index_rec s lim (i + 1) c
+
+let index s c = index_rec s (length s) 0 c
+
+let rec rindex_rec s i c =
+  if i < 0 then raise Not_found else
+  if unsafe_get s i = c then i else rindex_rec s (i - 1) c
+
+let rindex s c = rindex_rec s (length s - 1) c
+
+let index_from s i c =
+  let l = length s in
+  if i < 0 || i > l then invalid_arg "String.index_from / Bytes.index_from" else
+    index_rec s l i c
+
 let rindex_from s i c =
-  B.rindex_from (bos s) i c
-let contains s c =
-  B.contains (bos s) c
+  if i < -1 || i >= length s then
+    invalid_arg "String.rindex_from / Bytes.rindex_from"
```

Comment:
Good point, but that is a breaking change and we only rolled out the literal-strings-in-patterns warning last release. If I was release manager (I'm not), I would probably ask for the whole change to wait for 4.05, but would certainly not be enthusiastic about changing exception names at this point of the release.